File: Collections/BitArray.vb
Public Class BitArray
    Inherits BaseList(Of Boolean)

    Public Sub New(ByVal Count As Integer)
        Me._Count = Count
        Me._Bytes = New Byte((Count >> 3) + If((Count And 7) = 0, 0, 1) - 1) {}
    End Sub

    Public Overrides Sub Insert(index As Integer, item As Boolean)
        Throw New NotSupportedException()
    End Sub

    Public Overrides Sub RemoveAt(index As Integer)
        Throw New NotSupportedException()
    End Sub

    Public Overrides Sub Clear()
        Array.Clear(Me._Bytes, 0, Me._Bytes.Length)
    End Sub

    Private Function LastByteBitCount() As Integer
        Dim R = Me.Count And 7
        If R = 0 Then
            R = 8
        End If
        Return R
    End Function

    Public Sub [Not]()
        For I = 0 To Me._Bytes.Length - 1
            Me._Bytes(I) = Not Me._Bytes(I)
        Next
    End Sub

    ''' <param name="Amount">Shift left if positive, right if negative.</param>
    Public Sub Shift(ByVal Amount As Integer)
        For I = 0 To Me._Bytes.Length - 1
            Me._Bytes(I) = Not Me._Bytes(I)
        Next
        Me._Bytes(Me._Bytes.Length - 1) = CByte(Me._Bytes(Me._Bytes.Length - 1) And ((1 << Me.LastByteBitCount()) - 1))
    End Sub

    Protected Overrides Function IEnumerable_1_GetEnumerator() As IEnumerator(Of Boolean)
        Return Me.GetEnumerator()
    End Function

    Public Iterator Function GetEnumerator() As IEnumerator(Of Boolean)
        For I = 0 To Me._Bytes.Length - 2
            Dim B = Me._Bytes(I)
            For J = 0 To 7
                Yield ((B >> J) And 1) = 1
            Next
        Next

        If True Then
            Dim B = Me._Bytes(Me._Bytes.Length - 1)

            For J = 0 To Me.LastByteBitCount() - 1
                Yield ((B >> J) And 1) = 1
            Next
        End If
    End Function

    Public Sub SetOne(ByVal Index As Integer)
        Verify.True(0 <= Index And Index < Me.Count, "Index out of range.")

        Dim B As Integer = Me._Bytes(Index >> 3)
        Dim I = Index And 7
        B = B Or (1 << I)
        Me._Bytes(Index >> 3) = CByte(B)
    End Sub

    Public Sub SetZero(ByVal Index As Integer)
        Verify.True(0 <= Index And Index < Me.Count, "Index out of range.")

        Dim B As Integer = Me._Bytes(Index >> 3)
        Dim I = Index And 7
        B = B And Not (1 << I)
        Me._Bytes(Index >> 3) = CByte(B)
    End Sub

    Default Public Overrides Property Item(index As Integer) As Boolean
        Get
            Verify.True(0 <= index And index < Me.Count, "Index out of range.")

            Dim B = Me._Bytes(index >> 3)
            Dim I = index And 7
            Return ((B >> I) And 1) = 1
        End Get
        Set(value As Boolean)
            Verify.True(0 <= index And index < Me.Count, "Index out of range.")

            Dim B As Integer = Me._Bytes(index >> 3)
            Dim I = index And 7
            B = (B And Not (1 << I)) Or (If(value, 1, 0) << I)
            Me._Bytes(index >> 3) = CByte(B)
        End Set
    End Property

    Public Property [Byte](Index As Integer) As Byte
        Get
            Return Me._Bytes(Index)
        End Get
        Set(Value As Byte)
            If Index = Me._Bytes.Length - 1 Then
                Value = CByte(Value And ((1 << Me.LastByteBitCount()) - 1))
            End If
            Me._Bytes(Index) = Value
        End Set
    End Property

#Region "BytesCount Property"
    Public ReadOnly Property BytesCount As Integer
        Get
            Return Me._Bytes.Count
        End Get
    End Property
#End Region

#Region "Count Property"
    Private ReadOnly _Count As Integer

    Public Overrides ReadOnly Property Count As Integer
        Get
            Return Me._Count
        End Get
    End Property
#End Region

    Private _Bytes As Byte()

End Class

File: Collections/BitArray.cs
﻿using System.Linq;
using System.Collections.Generic;
using System;

namespace Ks
{
    namespace Common
    {
        public class BitArray : BaseList<bool>
        {
            public BitArray(int Count)
            {
                this._Count = Count;
                this._Bytes = new byte[((Count >> 3) + (((Count & 7) == 0) ? 0 : 1)) - 1 + 1];
                this.Byte = new IndexerProperty<BitArray, byte, int>(this, this.Byte.Getter, this.Byte.Setter);
            }

            public override void Insert(int index, bool item)
            {
                throw new NotSupportedException();
            }

            public override void RemoveAt(int index)
            {
                throw new NotSupportedException();
            }

            public override void Clear()
            {
                Array.Clear(this._Bytes, 0, this._Bytes.Length);
            }

            private int LastByteBitCount()
            {
                var R = this.Count & 7;
                if (R == 0)
                    R = 8;
                return R;
            }

            public void Not()
            {
                var loopTo = this._Bytes.Length - 1;
                for (var I = 0; I <= loopTo; I++)
                    this._Bytes[I] = (byte)~this._Bytes[I];
            }

            /// <param name="Amount">Shift left if positive, right if negative.</param>
            public void Shift(int Amount)
            {
                var loopTo = this._Bytes.Length - 1;
                for (var I = 0; I <= loopTo; I++)
                    this._Bytes[I] = (byte)~this._Bytes[I];
                this._Bytes[this._Bytes.Length - 1] = System.Convert.ToByte((this._Bytes[this._Bytes.Length - 1] & ((1 << this.LastByteBitCount()) - 1)));
            }

            protected override IEnumerator<bool> _GetEnumerator()
            {
                return this.GetEnumerator();
            }

            public IEnumerator<bool> GetEnumerator()
            {
                var loopTo = this._Bytes.Length - 2;
                for (var I = 0; I <= loopTo; I++)
                {
                    var B = this._Bytes[I];
                    for (var J = 0; J <= 7; J++)
                        yield return (B >> J & 1) == 1;
                }

                if (true)
                {
                    var B = this._Bytes[this._Bytes.Length - 1];
                    var loopTo1 = this.LastByteBitCount() - 1;
                    for (var J = 0; J <= loopTo1; J++)
                        yield return (B >> J & 1) == 1;
                }
            }

            public void SetOne(int Index)
            {
                Verify.True((0 <= Index) & (Index < this.Count), "Index out of range.");

                int B = (int)this._Bytes[Index >> 3];
                var I = Index & 7;
                B = B | (1 << I);
                this._Bytes[Index >> 3] = System.Convert.ToByte(B);
            }

            public void SetZero(int Index)
            {
                Verify.True((0 <= Index) & (Index < this.Count), "Index out of range.");

                int B = (int)this._Bytes[Index >> 3];
                var I = Index & 7;
                B = B & ~(1 << I);
                this._Bytes[Index >> 3] = System.Convert.ToByte(B);
            }

            public override bool this[int index]
            {
                get
                {
                    Verify.True((0 <= index) & (index < this.Count), "Index out of range.");

                    var B = this._Bytes[index >> 3];
                    var I = index & 7;
                    return (B >> I & 1) == 1;
                }
                set
                {
                    Verify.True((0 <= index) & (index < this.Count), "Index out of range.");

                    int B = (int)this._Bytes[index >> 3];
                    var I = index & 7;
                    B = (B & ~(1 << I)) | ((value ? 1 : 0) << I);
                    this._Bytes[index >> 3] = System.Convert.ToByte(B);
                }
            }

            public IndexerProperty<BitArray, byte, int> Byte { get; } = new IndexerProperty<BitArray, byte, int>(
                null,
                (self, Index) =>
                {
                    return self._Bytes[Index];
                },
                (self, Index, value) =>
                {
                    if (Index == (self._Bytes.Length - 1))
                        value = System.Convert.ToByte((value & ((1 << self.LastByteBitCount()) - 1)));
                    self._Bytes[Index] = value;
                }
            );

            public int BytesCount
            {
                get
                {
                    return this._Bytes.Count();
                }
            }

            private readonly int _Count;

            public override int Count
            {
                get
                {
                    return this._Count;
                }
            }

            private byte[] _Bytes;
        }
    }
}
